### ISSUE TITLE
App stuck in crash loop (beta apk)

### ISSUE THREAD

**Issue by collectorgeneral** [y20k/trackbook#7]
Coming from #5, I had tested the [beta apk](https://github.com/y20k/trackbook/issues/5#issuecomment-271994193). The total duration and all is fine now. However a strange issue occured where I was soft locked out of Trackbook. 

This issue is present most of the time. To reporduce the issue, just start tracking as usual, let the app do its thing in the background. Now sometime after, open the app to see your route. 

There are no points displayed on the map! HOWEVER, the notification is fine and is updated as usual. Also, once you STOP tracking, the points are visible on the map. You can then proceed to save them without issue.

However, before you hit stop, if you hit the 'Locate me' thingy, it instantly crashes! And what more, you cannot open the app without it crashing again! I had to force clear app cache and data for it to work normally. I have attached a filtered catlog. Let me know if there is anything missing here.
[tb_crash_locat_20170113.txt](https://github.com/y20k/trackbook/files/705096/tb_crash_locat_20170113.txt)

**Comment by collectorgeneral** [comment]
I can reproduce this issue consistently, I have attached a video showing the procedure in case you wanted to look at it.
[ScreenCast_TrackBook_20170113.zip](https://github.com/y20k/trackbook/files/705471/ScreenCast_TrackBook_20170113.zip)

**Comment by y20k** [comment]
@collectorgeneral: Thank you! That is **VERY** helpful. I will look into this strange behaviour and report back here.

**Comment by y20k** [comment]
@collectorgeneral: Thanks again for the video detailing the crash! The key failure seems to be the part, where one re-enters the app, and the red dot indicating the own location is missing.

I did try to recreate the error on my two Android devices. A Nexus 5 phone running Android 6 and a Pixel C tablet with Android 7. On both I did try the steps you showed in your video - with the System Setting "Don't keep activities" both on and off.

I specifically tried this option because it can simulate a state where the UI of Trackbook (MainActivity) gets killed by the system. In that case the app recreates its state from a so called saved instance. I put among other things the current recording and the last known location in that "saved instance". Observing the faulty behavior I think it has something to do with the saved state and the recreation of the UI.

Anyway. The main problem is: I could not recreate the crash. There is one explanation, that would be great i true. I did implement some changes on how the app restores its state. For example I did implement a temp-file representing a finished (but not yet saved) recording that gets created and loaded under certain conditions.

There is a chance, that if you just installed the last APK over an existing beta installation of Trackbook that the new version tries to recreate a "saved instance" that is incompatible.

**tldr;** -> My hope is: If you would just completely un-install  and then reinstall the app, the error might be gone. 

Please report back, if that changed anything. **If not,** I would need a longer error log. A log over a period of 20 seconds (or how long it takes to invoke the crash) - lines containing "org.y20k.trackbook" are of interest to me. Anything else you can strip out. I mean I can do that, too. But maybe for privacy reasons it is better if you would do it yourself. You can send me logs via mail, if you want (mail@y20k.org).

**Comment by collectorgeneral** [comment]
@y20k I thought the same too, so I did that before hand. Unfortunately the problem persists. I did it again just now, could reproduce it. You do not have to wait for any time, just start tracking and go back. I have attached a catlog for the event, I did it before and after removing and installing the application. The log should show you the events.
[2017-01-14-15-58-43.txt](https://github.com/y20k/trackbook/files/706185/2017-01-14-15-58-43.txt)

**Comment by y20k** [comment]
THX. I will study the log tomorrow and report any findings here.

**Comment by y20k** [comment]
@collectorgeneral I am looking through the log line by line. I am not sure what really happened on your device, yet. The background service seems to crash. Probably because it did not receive the last know location from the ui (= no red my location dot). On the other hand: There should be a fallback for that case. 

I need to study the log a bit more.

One thing you can do, if you find the time. There is still the possibility that the app was not uninstalled completely. Under certain circumstances an app gets installed for multiple local users. In that case a simple uninstall does not do the trick you have to chose **"Uninstall for all users"**.

![Uninstall for all users](https://cloud.githubusercontent.com/assets/9103935/21984489/7bfc40a0-dbf6-11e6-9fa7-2b383e9e9a87.jpg)

Also there is the possibility that the Android Backup service re-installs files created by the beta version that can cause complications. To make sure that does not happen. Install the new APK start the app, finish on-boarding (permissions). Wait a minute and then **"Clear data"**.

![Clear data](https://cloud.githubusercontent.com/assets/9103935/21984600/fba56ad4-dbf6-11e6-9c3e-1ff1678930ba.jpg)

In the meantime I will continue to try to find out what happened.

**Comment by collectorgeneral** [comment]
@y20k I do not have that option, I do not think it is installed for all users. Nevertheless, following the other steps led to the same result.

**Comment by y20k** [comment]
Okay. Good to know. Thanks.

**Comment by y20k** [comment]
One question: Could you scan your logs for the phrase **`Problem accessing or modifying the mScroller field.`** ?

**Comment by collectorgeneral** [comment]
You it exists. Its present in the second log I attached. For your reference:

01-14 15:45:54.085 E/NonSwipeableViewPager(14385): Problem accessing or modifying the mScroller field. 01-14 15:45:54.085 W/System.err(14385): java.lang.NoSuchFieldException: No field mScroller in class Landroid/support/v4/j/au; (declaration of 'android.support.v4.j.au' appears in /data/app/org.y20k.trackbook-1/base.apk) 01-14 15:45:54.086 W/System.err(14385): at java.lang.Class.getDeclaredField(Native Method) 01-14 15:45:54.086 W/System.err(14385): at org.y20k.trackbook.layout.NonSwipeableViewPager.f(Unknown Source) 01-14 15:45:54.086 W/System.err(14385): at org.y20k.trackbook.layout.NonSwipeableViewPager.<init>(Unknown Source) 01-14 15:45:54.086 W/System.err(14385): at java.lang.reflect.Constructor.newInstance(Native Method) 01-14 15:45:54.086 W/System.err(14385): at android.view.LayoutInflater.createView(LayoutInflater.java:619) 01-14 15:45:54.089 W/System.err(14385): at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:764) 01-14 15:45:54.089 W/System.err(14385): at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704) 01-14 15:45:54.089 W/System.err(14385): at android.view.LayoutInflater.rInflate(LayoutInflater.java:835) 01-14 15:45:54.089 W/System.err(14385): at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 01-14 15:45:54.089 W/System.err(14385): at android.view.LayoutInflater.inflate(LayoutInflater.java:515) 01-14 15:45:54.089 W/System.err(14385): at android.view.LayoutInflater.inflate(LayoutInflater.java:423) 01-14 15:45:54.089 W/System.err(14385): at android.view.LayoutInflater.inflate(LayoutInflater.java:374) 01-14 15:45:54.089 W/System.err(14385): at android.support.v7.app.k.b(Unknown Source) 01-14 15:45:54.089 W/System.err(14385): at android.support.v7.app.c.setContentView(Unknown Source) 01-14 15:45:54.089 W/System.err(14385): at org.y20k.trackbook.MainActivity.j(Unknown Source) 01-14 15:45:54.089 W/System.err(14385): at org.y20k.trackbook.MainActivity.onCreate(Unknown Source) 01-14 15:45:54.089 W/System.err(14385): at android.app.Activity.performCreate(Activity.java:6251) 01-14 15:45:54.089 W/System.err(14385): at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108) 01-14 15:45:54.089 W/System.err(14385): at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2403) 01-14 15:45:54.089 W/System.err(14385): at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2510) 01-14 15:45:54.089 W/System.err(14385): at android.app.ActivityThread.-wrap11(ActivityThread.java) 01-14 15:45:54.089 W/System.err(14385): at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1363) 01-14 15:45:54.089 W/System.err(14385): at android.os.Handler.dispatchMessage(Handler.java:102) 01-14 15:45:54.089 W/System.err(14385): at android.os.Looper.loop(Looper.java:148) 01-14 15:45:54.089 W/System.err(14385): at android.app.ActivityThread.main(ActivityThread.java:5461) 01-14 15:45:54.089 W/System.err(14385): at java.lang.reflect.Method.invoke(Native Method) 01-14 15:45:54.089 W/System.err(14385): at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 01-14 15:45:54.089 W/System.err(14385): at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 01-14 15:45:54.093 I/RecordingDevice(14323): Dequeued buffer 16664112704 01-14 15:45:54.113 I/RecordingDevice(14323): Dequeued buffer 16664129759

**Comment by y20k** [comment]
Okay. Thanks. 

If the custom **NonSwipeableViewPager** is the problem, I can implement a check, if the app is really able access the Scroller within the original **ViewPager**. I am not sure why that would cause a problem on Cyanogen devices.

Anyway the last [commit](https://github.com/y20k/trackbook/commit/93176e2a7acae8ec640eadf62706d208c816f99f) implements such a check. Can you test if that changes anything? Here a new APK [trackbook-app-test-1.0.0d.apk](https://www.dropbox.com/s/36y1dwd4zklnr8u/trackbook-app-test-1.0.0d.apk?dl=0)

Please check the logs for occurrences of either `Unable to get mScroller field.` or `Problem accessing or modifying the mScroller field. Exception: XYZ`.

**Comment by y20k** [comment]
PS. I have an alternative version of the APK that is compiled without the library **com.android.support:support-v4**.

[trackbook-app-test-1.0.0e-without-support-v4](https://www.dropbox.com/s/clsth4tupnh4jss/trackbook-app-test-1.0.0e-without-support-v4.apk?dl=0)

It is worth testing also. Please make sure that there is no old data present -> **Clear data**

Thanks for the continued testing! Too bad that I cannot reproduce the problem.

**Comment by collectorgeneral** [comment]
Unfortunately the logs still indicate the issue is present. I cleared data, uninstalled, installed from apk, cleared data after first launch and launched again. Surprisingly both cases gave me the issue.

01-16 12:33:55.748 W/System.err(21044): java.lang.NoSuchFieldException: No field mScroller in class Landroid/support/v4/j/au; (declaration of 'android.support.v4.j.au' appears in /data/app/org.y20k.trackbook-1/base.apk)

01-16 12:34:02.366 E/NonSwipeableViewPager(21044): Problem accessing or modifying the mScroller field. Exception: java.lang.NoSuchFieldException: No field mScroller in class Landroid/support/v4/j/au; (declaration of 'android.support.v4.j.au' appears in /data/app/org.y20k.trackbook-1/base.apk)

Are you sure you shared the correct apk (the without v4 support)? It gave me the exact same error:
01-16 12:40:40.416 E/NonSwipeableViewPager(24735): Problem accessing or modifying the mScroller field. Exception: java.lang.NoSuchFieldException: No field mScroller in class Landroid/support/v4/j/au; (declaration of 'android.support.v4.j.au' appears in /data/app/org.y20k.trackbook-1/base.apk)


On 16 January 2017 12:29:56 GMT-04:00, y20k <notifications@github.com> wrote:
>Okay. Thanks. 
>
>If the custom **NonSwipeableViewPager** is the problem, I can implement
>a check, if the app is really able access the Scroller within the
>original **ViewPager**. I am not sure why that would cause a problem on
>Cyanogen devices.
>
>Anyway the last
>[commit](https://github.com/y20k/trackbook/commit/93176e2a7acae8ec640eadf62706d208c816f99f)
>implements such a check. Can you test if that changes anything? Here a
>new APK
>[trackbook-app-test-1.0.0d.apk](https://www.dropbox.com/s/36y1dwd4zklnr8u/trackbook-app-test-1.0.0d.apk?dl=0)
>
>Please check the logs for occurrences of either `Unable to get
>mScroller field.` or `Problem accessing or modifying the mScroller
>field. Exception: XYZ`.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/y20k/trackbook/issues/7#issuecomment-272907425

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by y20k** [comment]
Okay. Let's try a version of the app that does not have a transition between tabs --> [commit ed0f0db](https://github.com/y20k/trackbook/commit/ed0f0db464213ff2191350b0444e74b26d976dc3). 

The whole troublesome portion of the code is not present. 

Here an updated APK: [trackbook-app-test-1.0.0f.apk](https://www.dropbox.com/s/q8k4j7m647sk7on/trackbook-app-test-1.0.0f.apk?dl=0)

**The big question is: Does the app still crash?**

**Comment by collectorgeneral** [comment]
Strangely, I did not find any mscroller references, and the app still crashed. So I am now wondering if we are looking at the wrong thing.

01-16 13:07:29.499 V/MainActivity(31073): MainActivity: CASE MY LOCATION. 
01-16 13:07:29.499 D/AndroidRuntime(31073): Shutting down VM 
01-16 13:07:29.501 E/AndroidRuntime(31073): FATAL EXCEPTION: main 
01-16 13:07:29.501 E/AndroidRuntime(31073): Process: org.y20k.trackbook, PID: 31073 
01-16 13:07:29.501 E/AndroidRuntime(31073): Theme: themes:{} 
01-16 13:07:29.501 E/AndroidRuntime(31073): java.lang.NullPointerException: Attempt to invoke virtual method 'int org.y20k.trackbook.a.a.c()' on a null object reference 
01-16 13:07:29.501 E/AndroidRuntime(31073): at org.y20k.trackbook.a.a(Unknown Source)
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v4.b.k.c(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v4.b.q.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v4.b.n.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v4.b.l.onMenuItemSelected(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.app.c.onMenuItemSelected(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.view.i.onMenuItemSelected(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.view.i.onMenuItemSelected(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.app.t$2.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.widget.Toolbar$1.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.widget.ActionMenuView$d.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.view.menu.h.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.view.menu.j.b(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.view.menu.h.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.view.menu.h.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.widget.ActionMenuView.a(Unknown Source) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.support.v7.view.menu.ActionMenuItemView.onClick(Unknown Source)
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.view.View.performClick(View.java:5204) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.view.View$PerformClick.run(View.java:21158) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.os.Handler.handleCallback(Handler.java:739) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.os.Handler.dispatchMessage(Handler.java:95) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.os.Looper.loop(Looper.java:148)
01-16 13:07:29.501 E/AndroidRuntime(31073): at android.app.ActivityThread.main(ActivityThread.java:5461) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at java.lang.reflect.Method.invoke(Native Method) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
01-16 13:07:29.501 E/AndroidRuntime(31073): at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
01-16 13:07:29.502 W/ActivityManager(5203): Force finishing activity org.y20k.trackbook/.MainActivity

**Comment by y20k** [comment]
> So I am now wondering if we are looking at the wrong thing.

Yep. Definitely seems to be the case. I need to think in a different direction.

**Comment by y20k** [comment]
Can you search your log again for the phrase `TrackingState:` (true or false). And can you paste a part of the log here. Maybe ten lines before and ten lines after the occurence?

**Comment by collectorgeneral** [comment]
I found only one instance of the tracking state. Here it is:

01-16 13:07:24.746 I/RecordingDevice(20789): Dequeued buffer 157369160160
01-16 13:07:24.772 I/RecordingDevice(20789): Dequeued buffer 157369227519 01-16 13:07:24.798 I/RecordingDevice(20789): Dequeued buffer 157369277045 01-16 13:07:24.827 I/RecordingDevice(20789): Dequeued buffer 157369360556 01-16 13:07:24.851 I/RecordingDevice(20789): Dequeued buffer 157369426758 01-16 13:07:24.876 I/RecordingDevice(20789): Dequeued buffer 157369493366 01-16 13:07:24.946 I/RecordingDevice(20789): Dequeued buffer 157369577308 01-16 13:07:25.008 I/RecordingDevice(20789): Dequeued buffer 157369626686 01-16 13:07:25.064 I/RecordingDevice(20789): Dequeued buffer 157369693733 01-16 13:07:25.130 I/RecordingDevice(20789): Dequeued buffer 157369760869 01-16 13:07:25.182 I/RecordingDevice(20789): Dequeued buffer 157369809850 01-16 13:07:25.188 V/a (31073): Removing location listeners. 01-16 13:07:25.189 V/a (31073): Removing gps listener. 01-16 13:07:25.192 V/a (31073): Removing network listener. 01-16 13:07:25.192 V/a (31073): Stopping preliminary tracking. 01-16 13:07:25.192 V/a (31073): TrackingState: true 01-16 13:07:25.207 W/Binder_1(5215): type=1400 audit(0.0:17816): avc: denied { ioctl } for path="socket:[3488799]" dev="sockfs" ino=3488799 ioctlcmd=7704 scontext=u:r:system_server:s0 tcontext=u:r:system_server:s0 tclass=unix_stream_socket permissive=0 01-16 13:07:25.207 W/Binder_1(5215): type=1400 audit(0.0:17817): avc: denied { ioctl } for path="socket:[3488799]" dev="sockfs" ino=3488799 ioctlcmd=7704 scontext=u:r:system_server:s0 tcontext=u:r:system_server:s0 tclass=unix_stream_socket permissive=0 01-16 13:07:25.207 V/TrackerService(31073): Service received command: START 01-16 13:07:25.208 I/RecordingDevice(20789): Dequeued buffer 157369827054 01-16 13:07:25.209 I/qti_sensors_hal(5203): Step Counter: batch 15.00 inf 01-16 13:07:25.209 I/qti_sensors_hal(5203): Step Counter: activate=1 01-16 13:07:25.236 I/RecordingDevice(20789): Dequeued buffer 157369843405 01-16 13:07:25.256 V/a (31073): Registering location listeners. 01-16 13:07:25.260 V/a (31073): Registering gps listener. 01-16 13:07:25.261 V/a (31073): Registering network listener. 01-16 13:07:25.265 I/RecordingDevice(20789): Dequeued buffer 157369860128 01-16 13:07:25.284 I/RecordingDevice(20789): Dequeued buffer 157369876340 01-16 13:07:25.305 V/a (31073): MapFragment: drawing new track overlay. 01-16 13:07:25.306 V/a (31073): MapFragment: Track update received - drawing map. 01-16 13:07:25.310 I/RecordingDevice(20789): Dequeued buffer 157369893311

**Comment by y20k** [comment]
Okay. Thanks. 

I have another idea of what could have caused the problem. The tabbed UI consists of three parts: A frame (MainActivity) and two maps (MainActivityMapFragment & MainActivityTrackFragment). From time to time the frame has to communicate to the maps. For instance when the user presses the "my location" button - or the "start/stop" button.

If you see `TrackingState:` in the logs that is a sign that this frame-to-map communication has worked. That is why I was asking. So at least in one case it was working.

The Android system can kill Activities and Fragments if the app is not visible. I think it is possible that the system only kills a Fragment but not its hosting Activity (the frame). In that case the Activity has an outdated reference to the old (now killed) fragment. The communication is broken.

I looked around for a possible fix and found a possible solution(*) for that problem.

I am not sure if that theoretical problem has any connection to your crash. Debugging without being able to reproduce a bug is a tedious endeavor. 

Anyway. I did implement some changes in the way the frame keeps references to its maps. Last [commit](https://github.com/y20k/trackbook/commit/ce452905fe4c04438c22d1ac5deae447fa5b3599) - and the signed APK for that codebase: [trackbook-app-test-1.0.0g.apk](https://www.dropbox.com/s/hadc2lh750ci3g1/trackbook-app-test-1.0.0g.apk?dl=0)

Want to try it out?

(*) [Android Activity To Fragment Communication](http://www.truiton.com/2015/12/android-activity-fragment-communication/)

**Comment by collectorgeneral** [comment]
No luck yet. I am really wondering what is going on here...

I pulled the log from my computer using adb, have a look at it to see if you can find something the other logs are missing.

[2017_01_17_TB_Log(2).txt](https://github.com/y20k/trackbook/files/710934/2017_01_17_TB_Log.2.txt)

**Comment by y20k** [comment]
Oh that log is easier to read! 

The last crash seems to have been caused by a tap on "My Location". 

I did implement a check for the case that the app thinks it is still recording. See [commit](https://github.com/y20k/trackbook/commit/6f7f58fa2399756bdf67e5df00ff064ac5ea5e7c). And there is a corresponding APK: ﻿[trackbook-app-test-1.0.0h.apk](https://www.dropbox.com/s/1izhcxhete8en0r/trackbook-app-test-1.0.0h.apk?dl=0)

**Comment by collectorgeneral** [comment]
FINALLY. No crash! 
I should have done this earlier, pulled the log through adb instead of my device, would have seen this WAY earlier. And the commit looks simple in hindsight as well. Oh wells. 

Now I am going to test it out for a bit more and try to break it. If I find anything I will post it here.

**Comment by y20k** [comment]
WOW. Nice. Thanks for sticking with me on that!

I think the detours I made until we found the error were not useless. The app is now better in many aspects. 

> If I find anything I will post it here.
👍

**Comment by collectorgeneral** [comment]
Alright, took a short walk around my block with the app active. 

- First, the infamous start and end time being same is back. The total duartion however, is calculated correctly this time around. 
`"c":983.90784,"d":735080,"e":1262.0,"f":"Jan 17, 2017 09:42:30","g":"Jan 17, 2017 09:42:30"}`

- Second, I broke the recording sessions into two sessions. I noticed that sometimes the recorded dots aren't visible when bringing the app to foreground again. The first time I could see it no issues, the second round however I couldn't see it. Now once you hit STOP, it becomes visible. You can then go about saving or clearing. I did not see any warnings or errors on my device logs. I will keep observing this behaviour.

**Comment by y20k** [comment]
To the latter:

This is easy to fix. I experience that too. I made the map request an update as soon as it becomes visible - at least as long the app is in recording mode. (see [commit](https://github.com/y20k/trackbook/commit/34125d71845df00613e9e73dcf1354216c1cd342))

And the date issue... I need to look into that.

**Comment by collectorgeneral** [comment]
Actually I just noticed that it takes a while (few seconds) after that it displays. So its not really a bug. I think you can ignore that for now!

**Comment by collectorgeneral** [comment]
Oh alright. Good to know!

**Comment by y20k** [comment]
Okay. The date bug. Saving the end date works on my phone. Maybe I thought it is the local device language that causes the problem. I tested UK-English, US-English and German. No problem on my device. (I guess I need you help again. 😞)

Next thing to do would be to trace the point where the correct date gets lost. That is why I added a few lines of extra logging (see [commit](https://github.com/y20k/trackbook/commit/28e9fb799185da3b167064a4f1fa5374e42908af)).

If you want to test... Check your logcat and look out for lines containing:

* "Track writing to parcel."
* "Track recreated from parcel."
* "Saving date and time to track:"

Here is an APK containing those extra logging instructions: [trackbook-app-test-1.0.0i.apk](https://www.dropbox.com/s/i8z9d2snjyqtx2c/trackbook-app-test-1.0.0i.apk?dl=0)

**Comment by collectorgeneral** [comment]
Roger that! I will go for a walk soon and will test it out and let you know.

**Comment by collectorgeneral** [comment]
I appreciate your effort and patience, unfortunately seems like we have some other issues to sort out 😭 .
 
I have been using it for a bit now and it seems to be crashing on startup. This happens when I leave it to track, leave it in the background for a while and open the app to stop and save recording. As soon as the app launches, it crashes. It happened to me a couple of times but I could launch it again to stop, save and clear. However, this time around the app opens, but I cannot stop anymore. Everytime I hit stop, it crashes. (FYI 'Locate me' does not crash it). 

I tried to dump the log using adb as much as fast as I can but I could not record the first crash. I filtered it by manually searching. Here is a huge log which hopefully captures all you need. The first block two or three 'blocks' indicate the instances where it instantly crashed after start. The last block is where it is stuck 'recording' and unable to stop without crashing. The last block repeats so I discarded it. If you feel there is something missing from the log, let me know. I will look it up in the huge log that I dumped and upload it if need be. 

[tb_crash_log_20170117_1935.txt](https://github.com/y20k/trackbook/files/712689/tb_crash_log_20170117_1935.txt)

**Comment by y20k** [comment]
Hi there.

Judging from the crash log there is a problem with saving and restoring dates. That could be related to language and time unit settings. I am guessing here. 

There is a way to force a certain format for a date when saving/restoring to a JSON text file. The change is just two lines of code (see [commit](https://github.com/y20k/trackbook/commit/d38adb62d95ed517c280320add6c003df83ca2f8)).

One problem though. That changes the format of your saved files. Older files will become unreadable. The app will crash. No problem yet, since nobody is using this version of the app. If you are trying out the new version below, please **delete all files within** `/Android/data/org.y20k.trackbook/files/tracks`.

So while I am at it: I added a version number to the format so that I hopefully can detect, if a file was written by an older version of Trackbook in the future. It is not unthinkable that the file format can change in future.versions. (see [commit](https://github.com/y20k/trackbook/commit/0e94bdb9dd93cf0e7ebcf24a006807443535333f)).

Anyway here is a new APK that works with the slightly changed file format: [trackbook-app-test-1.0.0ij.apk](https://www.dropbox.com/s/ued85chiz2rubdg/trackbook-app-test-1.0.0ij.apk?dl=0)

Okay. Let me again thank you for helping me here big time. These bugs are hard, because I cannot reproduce them. I am glad that you found them before I publish the new version. But I do not want repeat this over and over. Instead I added your GitHub handle to the **new [AUTHORS.md](https://github.com/y20k/trackbook/blob/master/AUTHORS.md) file** under a section "Testing". Also: I you ever come over to Stuttgart/Germany, I'll buy you a beer 🍻 .

**Comment by collectorgeneral** [comment]
@y20k I like testing and breaking stuff and love it when devs (such as yourself) actually take time to look and respond to them 😄 . 

Anyway I will do that, will let you know when I get more updates.

**Comment by collectorgeneral** [comment]
@y20k  Alright so I tested it out for a bit now, no crashes so far wohoo 🎉 !
The only thing left is the start and end date being the same. Did you remove the logging part of it? I cannot find any instances of parcel in the log 😆 . The total duration seems to be right but the start date is incorrect, it is equal to the end date. 
`...,"d":1133.4965,"e":885082,"f":1572.0,"g":"1/18/17 01:52 PM","h":"1/18/17 01:52 PM"}`

**Comment by y20k** [comment]
Hi @collectorgeneral, 

> but the start date is incorrect

Are you sure? I assumed that the end date was the one that is not correct. For that assumption I have an idea what could have went wrong.

It is a bit too late here to get started on that.

You could please confirm, if you think it the end date or the start date is the incorrect one?

In the meantime: Here is a screenshot showing that on my device start end end do differ. 😃 I feel compelled to post that screenshot - so to show anybody else reading this lengthy discussion that I am not an idiot.

![proof](https://cloud.githubusercontent.com/assets/9103935/22080477/08162766-ddc0-11e6-8574-13cfed3891cf.png)

**Comment by y20k** [comment]
And one last question for today. :-)

If you stop the recording, Trackbook writes a temp file, which is also just a JSON text file. The temp file gets deleted as soon as you actually "Save and Clear" or "Clear".

**Can you have a look at that temp file. Does it also contain the wrong date?** The temp file is also located in `/Android/data/org.y20k.trackbook/files/tracks`.

That would narrow down the areas where I would have to look.

**Comment by collectorgeneral** [comment]
Haha of course!

Yup, certain. The file saved in /adcard/Android/Data/org.y20k.trackbook is:

`2017-01-18-13-52-42.trackbook`

In the screen shot below you can see that I can travelled for 15 minutes:

![screenshot_20170118-161033](https://cloud.githubusercontent.com/assets/15972594/22081125/b698fcfa-dd98-11e6-8db9-2d8e663c97fe.png)

So the conclusion is that I the start date is bad.

**Comment by collectorgeneral** [comment]
Hmmm possibly not. Just did a quick check, the temp file indicates:

`"g":"1/18/17 04:17 PM","h":"1/18/17 04:18 PM"}`

So maybe it is somewhere after that the date gets weird. I will look into this a bit more 😄

**Comment by y20k** [comment]
One very last question for today. Could you please test the following:

1. Start the recording in portrait mode
2. Stop the recording via notification
3. Re-open the app in portrait mode
4. Switch to landscape mode
5. Save the track in landscape mode

Any recording longer than one minute should do the trick.Do you still have the same result (start = end)?

**Comment by collectorgeneral** [comment]
@y20k  Hm, no it shows different times. For the sake of information, I never had auto rotate enabled. 

![screenshot_20170118-171142](https://cloud.githubusercontent.com/assets/15972594/22083203/6371d26e-dda1-11e6-8a63-0ee2e343b226.png)

**Comment by collectorgeneral** [comment]
Strangely I haven't been able to reproduce it again. This time around it works perfectly fine. Not sure what's going on here. I will keep an eye out for it and see what's the deal.

**Comment by y20k** [comment]
Hi there. 

I am a bit time-restrained today. So I'll try to keep it short 😃.

I was always thinking that the difference between our devices was the way they handled memory. One device was killing the app earlier than the other. At first I thought it was the One Plus One - now I tend to think it is the Nexus 5.

If your device keeps the ui part of the app longer in memory, then there is the chance that the temp file never gets loaded - and the app tries to save a slightly outdated version of the recording - a version where the end has not been set. (FYI. A rotation change of the device forces a rebuild of the ui - and it did force the temp file to get loaded.)

What I can do (and did - see [commit](https://github.com/y20k/trackbook/commit/542660c0c556b7728e41582a81c415dfda62913a)) is to try to make sure that the temp file gets loaded no matter what state the ui is. It is actually good practice to do so. I wonder, why I did not think earlier of that.

Here is a signed APK reflecting the latest changes: [trackbook-app-test-1.0.0ik.apk](https://www.dropbox.com/s/yfwajqdtoeqch2q/trackbook-app-test-1.0.0ik.apk?dl=0)

That is probably all for today.

**Comment by collectorgeneral** [comment]
Roger that. I will keep a note of the behaviour today and post updates here.

**Comment by collectorgeneral** [comment]
Alright, I do not think we have the start time = end time issue sorted out. I have a bunch of sessions recorded today and all of them have the similar issue. 
I *think* I may have figured out a pattern to them (its a hypothesis, and may very well be wrong). In all of the sessions I realised that I tap the notification, open the app and then hit the stop button (which is what I always do, just to see the distance I travelled etc). I *think* that if you hit the stop button on the notification before proceeding to open the app and save, the dates are recorded correctly. Do you think that doing one way or the other affects what the app records? 

![screenshot_20170120-183330](https://cloud.githubusercontent.com/assets/15972594/22168112/f1f27200-df40-11e6-82d6-a164b99647a3.png)

**Comment by collectorgeneral** [comment]
I am starting to believe this is true. I just had another couple of sessions. The first sessions I tapped notification -> Stop recording, look at the result. Start time ==end time. 
Record another session->Stop from notification->Open application->save. This time proper start time and end time. Its probably worth investigating.

**Comment by y20k** [comment]
Not good, but: I can reproduce that behavior!

start recording -> switch to another app -> tap on notification (brings Trackbook in foreground) -> stop recording -> save recording  -> start time == end time

Problem is: Every part of the app can be killed at any time. So the background-service, the UI and the notification need to keep track of the correct state. 

Anyway. Since I can reproduce this, there should be an easy fix. THX

**Comment by y20k** [comment]
Okay. I think I got this fixed (see [commit](https://github.com/y20k/trackbook/commit/56ee80b1b4bbbb168096dbaf2abe344b76aaa95e)). I want to test the solution a bit. See if it did not introduce new problems.

I'll report back with a new APK soon.

**Comment by y20k** [comment]
This is a very long issue. It covers too many topics - some of them are resolved. I would like to close it. If something is still broken, please open a new issue. And in that case: Sorry for the inconvenience.
